### ISSUE TITLE
Bluez fixes for DualShock4 L2CAP problems

### ISSUE THREAD

**Issue by leonbottou** [MiSTer-devel/Linux_Image_creator_MiSTer#5]
# Initial problem
My DualShock4 game controller does not work with MiSTer (bluez-4.101) but works fine with my Linux laptop (bluez-5.48). The controller obviously implement the bluetooth protocol in a very spotty way, something that may be true from the real DS4 or may indicate that I own a cheap copy.

# Solution
It turns out that bluez5 contains patches to address some L2CAP MTU issues with the DS4 that seemed relevant to my symptoms. Backporting these patches to 4.101 works very well. The following diff contains my backporting patches: [ds4_l2cap_patch.txt](https://github.com/MiSTer-devel/Linux_Image_creator_MiSTer/files/4088892/ds4_l2cap_patch.txt)

# Compiling the patches.
This is tricky. I found that linux.img is based on buildroot 2017.02.  I read that Sorgelig had copied some Raspberry binaries and added his own to make the MiSTer linux rootfs. So I used buildroot 2017.02 to compile a rootfs that was derived from that of a raspberry3 but configured to look more like MiSTer (ARM A9, Linaro, eabihf, thumb2, udev, dbus libusb, libhid, bluez-4.101). Then setup the bluez package as an override source directory, with my patches. After compiling everything, I simply copied the resulting bluetoothd into the original linux.img.  Now my DS4 controller works.

**Comment by sorgelig** [comment]
Thanks for the patch!
Strange that for me DS4 works well with generic BT dongle.
I was about to make a new release of Linux, so your patch came in the last moment.

**Comment by sorgelig** [comment]
And it's good you've made the backport for bluez 4.x. 5.x has very different and complicated pairing procedure - that's why i don't want to move to 5.x

**Comment by leonbottou** [comment]
I am using a generic CSR dongle with this controller: https://www.amazon.com/gp/product/B07M9XWMWG.  

I do not have experience with other controllers. I did this because it wasn’t fun to buy another one (that’s cheating.)

 

Also  I have been playing with buildroot with the idea of seeing how to build a more reproducible linux.img (and maybe also build the kernel and u-boot as external gits).

Alas I went with Bluez5, but Bluez4 is still offerend in buldroot-2019.02.x.  Anyway, this does not look too hard.

 

Many thanks for your project. It’s fun.

 
Leon
